### ISSUE TITLE
Recommend using `it_behaves_like` vs `include_examples`

### ISSUE THREAD

**Issue by pirj** [rubocop-hq/rspec-style-guide#66]
The difference between the two is that the former creates a context, while the latter includes the examples and groups together with the context defined in the `shared_examples` that may lead to undesired consequences as overridden declarations.
I don't have good use cases for the latter in mind.

**Comment by matheussilvasantos** [comment]
In this case, I don't want to use the `it_behaves_like` terminology:

```ruby
describe Letter do
  describe "#send" do
    context "when it is a Monday" do
      let(:day_of_the_week) { 1 }

      include_examples "send letters"
    end

    context "when it is a Tuesday" do
      let(:day_of_the_week) { 2 }

      include_examples "send letters"
    end
  end
end

RSpec.shared_examples "send letters" do
  subject { described_class.new(day_of_the_week) }

  it "sends letters for the correct day" do    
    expect(subject.send).to eq(true)
  end
end
```

**Comment by pirj** [comment]
That is a specific case. Can you describe it in text?

**Comment by matheussilvasantos** [comment]
Sometimes you want to share examples among multiple tests, but the `it_behaves_like` terminology simply doesn't fit.
If the context behaviour for `include_examples` isn't helpful, perhaps it should be changed in RSpec to behave like `it_behaves_like`.

**Comment by pirj** [comment]
I meant something like

“Use include_examples if it’s the only such statement in the group, and no other groups are nested in the group that may be affected by a sudden definition of hooks, let variables, or methods”

If we settle on this, let’s see what bad and good examples we can provide.

Would you be up to harvest real-world-rspec for include_examples usages?

**Comment by matheussilvasantos** [comment]
Do you think the example here is good? https://github.com/rubocop/rspec-style-guide?tab=readme-ov-file#shared-examples

Also, is your recommendation
> If you want to use `include_examples` within a context, use only one shared example.

or

> If you want to use `include_examples` within a context, do not mix it with `it_behaves_like`.

I think neither are good.

If one wants to recommend something, I think the recommendation would be to not add `let` etc. within a shared example used as `include_examples` - I don't know if the lint is smart enough to catch that.

**Comment by matheussilvasantos** [comment]
In case it helps with a more realistic example, that's part of a spec I've written recently.

```ruby
context "when something" do
  context "and something else" do
    let(:date) { 3.days.ago }

    include_examples "returns only the single correct product"
  end

  context "and another thing" do
    let(:date) { 3.days.from_now }

    include_examples "returns only the single correct product"
  end
end
```

**Comment by pirj** [comment]
The lint may catch that, and we had an idea to have a cop that separates shared_context and shared_examples this way.

But this is specifically a style guide, and we shape the wording of the guidelines.

The example from shared_examples is good.

The reason for not using include_examples is obvious. The desire to use it - as well. We have to combine those into a clean concise paragraph.

**Comment by pirj** [comment]
It is worth mentioning that [RSpec Core's readme](https://github.com/rspec/rspec/blob/95fbf88255de04e4ba8288172292bda494c4f187/rspec-core/README.md?plain=1#L77) only mentions `include_examples`.


Yet another aspect: passing a block to it might not be the best idea. Both those examples will pass, and will both fail if you swap the order:
```ruby
    describe do
      shared_examples "equals foo" do
        it do
          expect(v).to eq("foo")
        end
      end

      include_examples "equals foo" do
        let(:v) { "bar" } # should 💥, but it won't
      end

      include_examples "equals foo" do
        let(:v) { "foo" }
      end
    end
```
Real-world [usages](https://github.com/search?q=%2Finclude_examples.%2B+do%2F+lang%3Aruby+path%3A%2Fspec%5C%2F%2F&type=code): [1](https://github.com/rubycdp/ferrum/blob/180f29297f2be7b770b32d8a2f73343e9530e67c/spec/browser_spec.rb#L418), [2](https://github.com/Mange/roadie-rails/blob/117b2567ad6aa4cbf1d1fc0a470f4f0ac756e2db/spec/integration_spec.rb#L56), [3](https://github.com/rubycdp/ferrum/blob/180f29297f2be7b770b32d8a2f73343e9530e67c/spec/browser_spec.rb#L418), [4](https://github.com/Mange/roadie-rails/blob/117b2567ad6aa4cbf1d1fc0a470f4f0ac756e2db/spec/integration_spec.rb#L56) etc.
I'll submit an issue or a PR to rspec-core to issue warnings on such usages.

Unexpected side effects cast a shade on its usage aesthetic.

**Comment by pirj** [comment]
This reads pretty well, don't you think?

```ruby
    context "when it is a Tuesday" do
      it_behaves_like "sends letters"
    end
```

```ruby
    it_behaves_like "it returns only the single correct product"
```

```ruby
    it_behaves_like "it equals foo"
```

I glanced over other [usages](https://github.com/search?q=%2Finclude_examples.%2B+do%2F+lang%3Aruby+path%3A%2Fspec%5C%2F%2F&type=code), but could not quickly find examples that do read well with "include examples", but wouldn't read well with "it behaves like".

You can also [alias](https://github.com/rspec/rspec/blob/12effb26900129ba2c977585ade9938d18f753ce/rspec-core/lib/rspec/core/configuration.rb#L1226C1-L1228C14) `it_behaves_like`:
```ruby
config.alias_it_behaves_like_to(:it_has_behavior, 'has behavior:')
```

**Comment by matheussilvasantos** [comment]
Personally, I don't think it reads well 🤷‍♂️

**Comment by pirj** [comment]
> [In RSpec, tests are not just scripts that verify your application code. They’re also specifications (or specs, for short): detailed explanations of how the application is supposed to behave, expressed in plain English.](https://github.com/rspec/rspec-rails/blob/a313574b17f408a365a62bbedd99f59f80b5ea03/README.md?plain=1#L6)

Let's use the above criteria and resort to the opinion of an unbiased third party.

<img width="782" height="639" alt="Image" src="https://github.com/user-attachments/assets/ab3444cb-8f29-4e7b-bb5d-fc804015dfb2" />

<img width="747" height="223" alt="Image" src="https://github.com/user-attachments/assets/142614df-4128-40c0-a8b4-7ad39444781c" />

:shrug:

**Comment by pirj** [comment]
Yet another case that can create problems as per [RSpec documentation](https://rspec.info/features/3-13/rspec-core/example-groups/shared-examples/):

> When you include parameterized examples in the current context multiple times, you may override previous method definitions and last declaration wins.

Indeed, even a block is not necessary for `include_examples` to leak context.

**Comment by matheussilvasantos** [comment]
I think `it_behaves_like` suits when you're testing sub-classes. Like:
- smartphone behaves like a computer
- smartphone behaves like a phone

But I don't feel like using it when I just want to run multiple assertions on multiple scenarios.

Sorry if I'm somewhat repeating myself.
